server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/TagAggregationOperator.java

```diff
@@ -0,0 +1,237 @@
+    return operatorContext;
+  }
+
+  @Override
+  public TsBlock next() {
+    long maxRuntime = operatorContext.getMaxRunTime().roundTo(TimeUnit.NANOSECONDS);
+    long start = System.nanoTime();
+    boolean successful = true;
+    while (System.nanoTime() - start < maxRuntime && !tsBlockBuilder.isFull() && successful) {
+      successful = processOneRow();
+    }
+    TsBlock tsBlock = null;
+    if (tsBlockBuilder.getPositionCount() > 0) {
+      tsBlock = tsBlockBuilder.build();
+    }
+    tsBlockBuilder.reset();
+    return tsBlock;
+  }
+
+  private boolean processOneRow() {
+    for (int i = 0; i < children.size(); i++) {
+      if (dataReady.isMarked(i)) {
+        continue;
+      }
+      if (dataUnavailable(i)) {
+        // Find next non-empty tsblock of the child
+        do {
+          inputTsBlocks[i] = children.get(i).next();
+        } while (inputTsBlocks[i] != null && inputTsBlocks[i].isEmpty());
+        consumedIndices[i] = 0;
+      }
+
+      // Blocked by children i
+      if (inputTsBlocks[i] == null) {
+        return false;
+      }
+      dataReady.mark(i);
+    }
+
+    TsBlock[] rowBlocks = new TsBlock[children.size()];
+    for (int i = 0; i < children.size(); i++) {
+      rowBlocks[i] = inputTsBlocks[i].getRegion(consumedIndices[i], 1);
+    }
+    for (int groupIdx = 0; groupIdx < groups.size(); groupIdx++) {
+      List<String> group = groups.get(groupIdx);
+      List<Aggregator> aggregators = groupedAggregators.get(groupIdx);
+
+      for (Aggregator aggregator : aggregators) {
+        if (aggregator == null) {
+          continue;
+        }
+        aggregator.reset();
+        aggregator.processTsBlocks(rowBlocks);
+      }
+
+      TimeColumnBuilder timeColumnBuilder = tsBlockBuilder.getTimeColumnBuilder();
+      timeColumnBuilder.writeLong(rowBlocks[0].getStartTime());
+      ColumnBuilder[] columnBuilders = tsBlockBuilder.getValueColumnBuilders();
+
+      for (int i = 0; i < group.size(); i++) {
+        if (group.get(i) == null) {
+          columnBuilders[i].writeBinary(new Binary("NULL"));
+        } else {
+          columnBuilders[i].writeBinary(new Binary(group.get(i)));
+        }
+      }
+      for (int i = 0; i < aggregators.size(); i++) {
+        Aggregator aggregator = aggregators.get(i);
+        ColumnBuilder columnBuilder = columnBuilders[i + group.size()];
+        if (aggregator == null) {
+          columnBuilder.appendNull();
+        } else {
+          aggregator.outputResult(new ColumnBuilder[] {columnBuilder});
+        }
+      }
+      tsBlockBuilder.declarePosition();
+    }
+
+    // Reset dataReady for next iteration
+    for (int i = 0; i < children.size(); i++) {
+      consumedIndices[i]++;
+      dataReady.unmark(i);
+    }
+    return true;
+  }
+
+  @Override
+  public boolean hasNext() {
+    for (int i = 0; i < children.size(); i++) {
+      if (dataUnavailable(i) && !children.get(i).hasNext()) {
+        return false;
+      }
```

Comment:
Only one child finised doesn't mean that other children are all finished.

Reply:
As we discussed in Slack, we work each child's output with a 'join' operation as the 1st level aggregation must be with the same segment rule, which results to the same segment output. So if one child is not ready, we treat that the whole operator tree is not ready.